compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -16504,6 +16522,28 @@ private STNode getArrayTypeDesc(STNode openBracket, STNode member, STNode closeB
         return lhsTypeDesc;
     }
 
+    private STNode getOptionalTypeDesc(STNode typeDescNode, STNode questionMarkToken) {
+        if (typeDescNode.kind == SyntaxKind.UNION_TYPE_DESC) {
+            STUnionTypeDescriptorNode unionTypeDesc = (STUnionTypeDescriptorNode) typeDescNode;
+            STNode middleTypeDesc = getOptionalTypeDesc(unionTypeDesc.rightTypeDesc, questionMarkToken);
+            typeDescNode = mergeLhsTypeAndRhsType(unionTypeDesc.leftTypeDesc, unionTypeDesc.pipeToken, middleTypeDesc);
+        } else if (typeDescNode.kind == SyntaxKind.INTERSECTION_TYPE_DESC) {
+            STIntersectionTypeDescriptorNode intersectionTypeDesc = (STIntersectionTypeDescriptorNode) typeDescNode;
+            STNode middleTypeDesc = getOptionalTypeDesc(intersectionTypeDesc.rightTypeDesc, questionMarkToken);
+            typeDescNode = mergeLhsTypeAndRhsType(intersectionTypeDesc.leftTypeDesc,
+                    intersectionTypeDesc.bitwiseAndToken, middleTypeDesc);
+        } else {
+            typeDescNode = createOptionalTypeDesc(typeDescNode, questionMarkToken);
+        }
+
+        return typeDescNode;
+    }
+
+    private STNode createOptionalTypeDesc(STNode typeDescriptorNode, STNode questionMarkToken) {
+        typeDescriptorNode = validateForUsageOfVar(typeDescriptorNode);
+        return STNodeFactory.createOptionalTypeDescriptorNode(typeDescriptorNode, questionMarkToken);
+    }
```

Comment:
We don't need a separate method right?. Let's include this in `getOptionalTypeDesc` methods itself and change the name of `getOptionalTypeDesc` to `createOptionalTypeDesc`